pkg/proxy/iptables/proxier.go

```diff
@@ -1025,10 +1025,17 @@ func (proxier *Proxier) syncProxyRules() {
 		// 2. ServiceTopology is not enabled.
 		// 3. EndpointSlice is not enabled (service topology depends on endpoint slice
 		// to get topology information).
-		if !svcInfo.OnlyNodeLocalEndpoints() && utilfeature.DefaultFeatureGate.Enabled(features.ServiceTopology) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceProxying) {
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceTopology) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceProxying) {
 			allEndpoints = proxy.FilterTopologyEndpoint(proxier.nodeLabels, svcInfo.TopologyKeys(), allEndpoints)
 		}
 
+		// Service InternalTrafficPolicy is only enabled when all of the
+		// following are true:
+		// 1. InternalTrafficPolicy is Local
+		// 2. ServiceInternalTrafficPolicy feature gate is on
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) && svcInfo.OnlyNodeLocalEndpointsForInternal() {
+			allEndpoints = proxy.FilterLocalEndpoint(svcInfo.InternalTrafficPolicy(), allEndpoints)
+		}
```

Comment:
One of the things I'm doing in the topology PR is pushing all of this logic into proxy so it's not repeated in each proxier. That would turn this into a call like this:

```
allEndpoints = proxy.FilterEndpoints(allEndpoints, svcInfo, proxier.nodeLabels)
```

That function then determines how to filter endpoints based on a combination of `svcInfo` and feature gates. I don't think that's strictly required for this PR, but wherever possible, I prefer pulling shared logic out of proxier implementations and into shared code.
https://github.com/kubernetes/kubernetes/pull/99522/files#diff-2ccafd63d57233fd0697b9852514604ab880e29aeb073035765ad7541cad1b35R1027

Reply:
This is definitely better and cleaner! I'll defer this change to your PR.